apps/rule-manager/app/service/RuleManager.scala

```diff
@@ -416,33 +416,55 @@ object RuleManager extends Loggable {
 
   def destructivelyPublishDictionaryRules(
       words: List[String],
-      bucketRuleResource: BucketRuleResource
+      bucketRuleResource: BucketRuleResource,
+      wordsToNotPublish: List[WordTag]
   ) = {
     // Destroy existing draft dictionary rules
     DbRuleDraft.destroyDictionaryRules()
     // Destroy existing live dictionary rules
     DbRuleLive.destroyDictionaryRules()
+    var availableTags = Tags.findAll()
+    // Add any tags from wordsToNotPublish that aren't already in the DB
+    val tagsToAdd = wordsToNotPublish
+      .map(wordTag => wordTag.tag)
+      .distinct
+      .filter(tagName => !availableTags.exists(tag => tag.name == tagName))
+      .map(name => Tag(None, name))
+    if (tagsToAdd.size > 0) {
+      Tags.batchInsert(tagsToAdd)
+      availableTags = Tags.findAll()
+    }
 
     val initialRuleOrder = DbRuleDraft.getLatestRuleOrder() + 1
-    val dictionaryRules = words.zipWithIndex.map(wordAndIndex =>
-      DbRuleDraft.withUser(
-        id = None,
-        ruleType = "dictionary",
-        pattern = Some(wordAndIndex._1),
-        category = Some("Collins Dictionary"),
-        ignore = false,
-        user = "Collins Dictionary",
-        ruleOrder = initialRuleOrder + wordAndIndex._2,
-        externalId = None
+    val dictionaryRules = words.zipWithIndex
+      .map(word => {
+        val tagId = wordsToNotPublish.find(taggedWord => taggedWord.word == word) match {
+          case Some(wordTag) => List(availableTags.find(tag => tag.name == wordTag.tag).get.id.get)
+          case _             => Nil
+        }
+        (word._1, word._2, tagId)
+      })
+      .map(wordIndexAndTag =>
+        DbRuleDraft.withUser(
+          id = None,
+          ruleType = "dictionary",
+          pattern = Some(wordIndexAndTag._1),
+          category = Some("Collins Dictionary"),
+          ignore = false,
+          user = "Collins Dictionary",
+          ruleOrder = initialRuleOrder + wordIndexAndTag._2,
+          externalId = None,
+          tags = wordIndexAndTag._3
+        )
       )
```

Comment:
A few things –
 - Do we need two maps here?
 - I'd be tempted to destructure the tuple to make it clear which bit is which.
 - Underscore notation makes things more concise, but you may prefer a more explicit syntax
 - We could use `flatMap` and `flatten` rather than `.get` to avoid unhappy paths throwing errors (although perhaps it's better to fail fast and loudly in this case!)

e.g.

```scala
    val dictionaryRules = words.zipWithIndex
      .map { case (word, index) =>
        val tagId = wordsToNotPublish.find(_.word == word) match {
          case Some(wordTag) =>
            List(availableTags.find(_.name == wordTag.tag).flatMap(_.id)).flatten
          case _ => Nil
        }
        DbRuleDraft.withUser(
          id = None,
          ruleType = "dictionary",
          pattern = Some(word),
          category = Some("Collins Dictionary"),
          ignore = false,
          user = "Collins Dictionary",
          ruleOrder = initialRuleOrder + index,
          externalId = None,
          tags = tagId
        )
      }
```

Reply:
All good suggestions, will include!